### PR TITLE
ENH: improve new seq / alignment moltype interconversion

### PR DESCRIPTION
**Pull request by GavinHuttley**
## Summary by Sourcery

Enhance the sequence and alignment modules by introducing a method for converting sequence arrays between different alphabets, and refactor dtype determination logic. Update tests to cover new functionality and exception handling changes.

Enhancements:
- Add a method to convert sequence arrays between different alphabets, improving interconversion capabilities.
- Refactor the logic for determining the smallest numpy integer dtype to include boundary values.

Tests:
- Add tests for the new sequence array conversion method, including validation checks for input and output sequences.
- Update tests to reflect changes in exception handling, replacing ValueError with more specific AlphabetError and MolTypeError.

### CODE REVIEW

File: tests/test_core/test_new_alphabet.py [hunk 1]
```diff
@@ -740,3 +740,17 @@ def test_get_subset_invalid():
...
+def test_convert_seq_array_to_check_valid():
+    bytes_alpha = new_moltype.BYTES.most_degen_alphabet()
+    dna_alpha = new_moltype.DNA.most_degen_alphabet()
+    # seq is invalid for DNA
+    seq = numpy.array([65, 67, 71, 84], dtype=numpy.uint8)
+    with pytest.raises(new_alphabet.AlphabetError):
+        dna_alpha.convert_seq_array_to(alphabet=bytes_alpha, seq=seq, check_valid=True)
+
+    # bytes alphabet seq with these low ordinals invalid for DNA output
+    seq = numpy.array([0, 1, 2, 3], dtype=numpy.uint8)
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply tests/test_core/test_new_alphabet.py:754]: many of those are invalid

File: tests/test_core/test_new_alignment.py [hunk 2]
```diff
@@ -2081,10 +2081,14 @@ def test_sequence_collection_to_moltype_with_gaps(mk_cls):
...
-    with pytest.raises(new_moltype.MolTypeError):
-        seqs.to_moltype("rna")
+    assert rna_seqs.to_dict() != dna_seqs.to_dict()
+
+    # and reverse this back to text
+    text_seqs = rna_seqs.to_moltype("text")
+    assert text_seqs.to_dict() == rna_seqs.to_dict()
+    # but this does not equal the original text seqs
+    # because of changing T to U
+    assert text_seqs.to_dict() != seqs.to_dict()
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply tests/test_core/test_new_alignment.py:2091]: that is done elsewhere
